File: aecp-python/src/aecp/recalibration.py
"""Score recalibration: map post-migration scores to ceiling-equivalent scores.

After embedding-space mapping, similarity scores are systematically compressed
(see Q10 study: margins shrink 39%, top-1 scores drop ~0.036). This breaks
downstream absolute-score thresholds silently.

ScoreRecalibrator fits an isotonic regression on holdout (mapped_score, ceiling_score)
pairs and provides a monotone map that restores score interpretability.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

import numpy as np


@dataclass
class RecalibrationReport:
    """Diagnostics for the fitted recalibrator."""

    n_pairs: int
    mean_mapped_score: float
    mean_ceiling_score: float
    mean_shift: float
    pre_margin_mean: float
    post_margin_mean: float
    margin_ratio: float
    threshold_agreement: dict[float, float] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        return {
            "n_pairs": self.n_pairs,
            "mean_mapped_score": self.mean_mapped_score,
            "mean_ceiling_score": self.mean_ceiling_score,
            "mean_shift": self.mean_shift,
            "pre_margin_mean": self.pre_margin_mean,
            "post_margin_mean": self.post_margin_mean,
            "margin_ratio": self.margin_ratio,
            "threshold_agreement": {str(k): v for k, v in self.threshold_agreement.items()},
        }


class ScoreRecalibrator:
    """Monotone map from mapped-space similarity scores to ceiling-equivalent scores.

    Uses isotonic regression trained on holdout query-document similarity pairs.

    Usage::

        # During calibration (automatic):
        recal = ScoreRecalibrator.fit_from_holdout(
            mapped_scores, ceiling_scores, query_ids, doc_ids, qrels
        )

        # At search time:
        calibrated = recal.transform(raw_scores)

        # Serialize:
        recal.save_dict()  # → stored in .aecp header under "score_recal_v1"
        recal2 = ScoreRecalibrator.from_dict(data)
    """

    def __init__(self) -> None:
        self._thresholds: np.ndarray | None = None
        self._values: np.ndarray | None = None
        self._report: RecalibrationReport | None = None

    @property
    def is_fitted(self) -> bool:
        return self._thresholds is not None

    @property
    def report(self) -> RecalibrationReport | None:
        return self._report

    def fit(
        self,
        mapped_scores: np.ndarray,
        ceiling_scores: np.ndarray,
    ) -> ScoreRecalibrator:
        """Fit isotonic regression on (mapped_score → ceiling_score) pairs.

        Parameters
        ----------
        mapped_scores:
            Similarity scores from the mapped retrieval (n_pairs,).
        ceiling_scores:
            Corresponding similarity scores from ceiling retrieval (n_pairs,).
        """
        from sklearn.isotonic import IsotonicRegression

        mapped_scores = np.asarray(mapped_scores, dtype=np.float64).ravel()
        ceiling_scores = np.asarray(ceiling_scores, dtype=np.float64).ravel()

        if len(mapped_scores) < 10:
            raise ValueError(f"Need ≥10 score pairs, got {len(mapped_scores)}")

        # Fit isotonic regression (increasing=True: higher mapped → higher ceiling)
        iso = IsotonicRegression(increasing=True, out_of_bounds="clip")
        iso.fit(mapped_scores, ceiling_scores)

        self._thresholds = iso.X_thresholds_.copy()
        self._values = iso.y_thresholds_.copy()

        # Compute report
        pre = mapped_scores
        post = iso.predict(mapped_scores)

        # Margin computation (sorted pairs, consecutive differences)
        pre_sorted = np.sort(pre)
        post_sorted = np.sort(post)
        pre_margin = float(np.mean(np.diff(pre_sorted[-20:])))  # top-20 margin
        post_margin = float(np.mean(np.diff(post_sorted[-20:])))

        # Threshold agreement
        thresholds = [0.5, 0.6, 0.7, 0.8, 0.9]
        agreement = {}
        for tau in thresholds:
            pre_decisions = pre >= tau
            post_decisions = post >= tau
            ceiling_decisions = ceiling_scores >= tau
            # Agreement = fraction where recalibrated matches ceiling
            agree = np.mean(post_decisions == ceiling_decisions)
            agreement[tau] = float(agree)

        self._report = RecalibrationReport(
            n_pairs=len(mapped_scores),
            mean_mapped_score=float(np.mean(mapped_scores)),
            mean_ceiling_score=float(np.mean(ceiling_scores)),
            mean_shift=float(np.mean(post) - np.mean(mapped_scores)),
            pre_margin_mean=pre_margin,
            post_margin_mean=post_margin,
            margin_ratio=post_margin / pre_margin if pre_margin > 1e-12 else 1.0,
            threshold_agreement=agreement,
        )

        return self

    def transform(self, scores: np.ndarray) -> np.ndarray:
        """Recalibrate scores: mapped → ceiling-equivalent."""
        if not self.is_fitted:
            raise RuntimeError("ScoreRecalibrator not fitted; call fit() first")
        scores = np.asarray(scores, dtype=np.float64)
        # Isotonic interpolation using stored breakpoints
        return np.interp(scores, self._thresholds, self._values)

    @classmethod
    def fit_from_holdout(
        cls,
        doc_src: np.ndarray,
        doc_tgt: np.ndarray,
        qry_tgt: np.ndarray,
        doc_ids: list[str],
        query_ids: list[str],
        qrels: dict[str, set[str]],
        mapped_docs: np.ndarray,
        n_random_pairs: int = 50_000,
        seed: int = 0,
    ) -> ScoreRecalibrator:
        """Build recalibrator from holdout data.

        Samples random query-document pairs + all top-20 neighbors to cover
        the full score range.
        """
        from aecp.mapping.base import l2_normalize

        rng = np.random.default_rng(seed)
        n_docs = len(doc_ids)
        n_queries = len(query_ids)

        # Normalize
        qt_n = l2_normalize(qry_tgt)
        dt_n = l2_normalize(doc_tgt)
        dm_n = l2_normalize(mapped_docs)

        # Ceiling similarities: qry × doc_tgt
        ceil_sims = qt_n @ dt_n.T  # (n_queries, n_docs)
        # Mapped similarities: qry × doc_mapped
        map_sims = qt_n @ dm_n.T  # (n_queries, n_docs)

        mapped_pairs = []
        ceiling_pairs = []

        # Sample random pairs (cover the low-score region)
        n_random = min(n_random_pairs, n_queries * n_docs)
        rand_q = rng.integers(0, n_queries, size=n_random)
        rand_d = rng.integers(0, n_docs, size=n_random)
        for qi, di in zip(rand_q, rand_d):
            mapped_pairs.append(float(map_sims[qi, di]))
            ceiling_pairs.append(float(ceil_sims[qi, di]))

        # Add all top-20 neighbor pairs (cover the high-score region)
        top_k = min(20, n_docs)
        for qi in range(n_queries):
            top_d = np.argsort(-map_sims[qi])[:top_k]
            for di in top_d:
                mapped_pairs.append(float(map_sims[qi, di]))
                ceiling_pairs.append(float(ceil_sims[qi, di]))

        recal = cls()
        recal.fit(np.array(mapped_pairs), np.array(ceiling_pairs))
        return recal

    def save_dict(self) -> dict[str, Any]:
        """Serialize to dict for .aecp header storage."""
        if not self.is_fitted:
            raise RuntimeError("Cannot serialize unfitted recalibrator")
        return {
            "thresholds": self._thresholds.tolist(),
            "values": self._values.tolist(),
            "report": self._report.to_dict() if self._report else None,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> ScoreRecalibrator:
        """Load from serialized dict."""
        recal = cls()
        recal._thresholds = np.array(data["thresholds"], dtype=np.float64)
        recal._values = np.array(data["values"], dtype=np.float64)
        if "report" in data and data["report"] is not None:
            recal._report = RecalibrationReport(**data["report"])
        return recal

    def save(self, path: str | Path) -> None:
        """Save recalibrator to JSON file."""
        path = Path(path)
        path.write_text(json.dumps(self.save_dict(), indent=2), encoding="utf-8")

    @classmethod
    def load(cls, path: str | Path) -> ScoreRecalibrator:
        """Load from JSON file."""
        data = json.loads(Path(path).read_text(encoding="utf-8"))
        return cls.from_dict(data)

File: isotrieve-npm/packages/core/src/recalibration.ts
/**
 * Score recalibration: map post-migration scores to ceiling-equivalent scores.
 *
 * After embedding-space mapping, similarity scores are systematically compressed.
 * ScoreRecalibrator fits an isotonic regression on holdout
 * (mapped_score, ceiling_score) pairs and provides a monotone map that restores
 * score interpretability.
 */

import type { RecalibrationReport, ScoreRecalibratorData } from './types';
import { l2Normalize } from './math/normalize';

export class ScoreRecalibrator {
  private _thresholds: Float64Array | null = null;
  private _values: Float64Array | null = null;
  private _report: RecalibrationReport | null = null;

  get isFitted(): boolean {
    return this._thresholds !== null;
  }

  get report(): RecalibrationReport | null {
    return this._report;
  }

  /**
   * Fit isotonic regression on (mapped_score -> ceiling_score) pairs.
   *
   * Isotonic regression: find the monotone non-decreasing function f that
   * minimizes sum of squared residuals. Uses the Pool Adjacent Violators
   * algorithm (PAVA).
   */
  fit(
    mappedScores: Float64Array | number[],
    ceilingScores: Float64Array | number[],
  ): this {
    const m = Float64Array.from(mappedScores);
    const c = Float64Array.from(ceilingScores);

    if (m.length < 10) {
      throw new Error(`Need >=10 score pairs, got ${m.length}`);
    }

    // Sort by mapped score
    const indices = Array.from({ length: m.length }, (_, i) => i);
    indices.sort((a, b) => m[a] - m[b]);
    const sortedM = new Float64Array(indices.map((i) => m[i]));
    const sortedC = new Float64Array(indices.map((i) => c[i]));

    // PAVA (Pool Adjacent Violators) for isotonic regression
    const n = sortedM.length;
    const fitted = new Float64Array(n);
    const blockStart = new Float64Array(n);
    const blockSize = new Float64Array(n);
    let nBlocks = 1;

    blockStart[0] = 0;
    blockSize[0] = 1;
    fitted[0] = sortedC[0];

    for (let i = 1; i < n; i++) {
      // Add new block
      blockStart[nBlocks] = i;
      blockSize[nBlocks] = 1;
      fitted[nBlocks] = sortedC[i];
      nBlocks++;

      // Merge blocks that violate monotonicity
      while (nBlocks > 1 && fitted[nBlocks - 2] > fitted[nBlocks - 1]) {
        const b2 = nBlocks - 1;
        const b1 = nBlocks - 2;
        const size1 = blockSize[b1];
        const size2 = blockSize[b2];
        const totalSize = size1 + size2;
        const mergedVal = (fitted[b1] * size1 + fitted[b2] * size2) / totalSize;

        fitted[b1] = mergedVal;
        blockSize[b1] = totalSize;
        nBlocks--;
      }
    }

    // Extract breakpoints (unique fitted values)
    const breakpoints: number[] = [];
    const breakValues: number[] = [];
    let prevVal = fitted[0] - 1;
    for (let i = 0; i < nBlocks; i++) {
      if (Math.abs(fitted[i] - prevVal) > 1e-15) {
        breakpoints.push(sortedM[Math.round(blockStart[i])]);
        breakValues.push(fitted[i]);
        prevVal = fitted[i];
      }
    }

    this._thresholds = new Float64Array(breakpoints);
    this._values = new Float64Array(breakValues);

    // Compute report
    const pre = m;
    const post = this.transform(pre);

    // Margin computation (top-20 sorted differences)
    const preSorted = Float64Array.from(pre).sort();
    const postSorted = Float64Array.from(post).sort();
    const marginSize = Math.min(20, preSorted.length);
    let preMargin = 0;
    let postMargin = 0;
    for (let i = preSorted.length - marginSize; i < preSorted.length - 1; i++) {
      preMargin += preSorted[i + 1] - preSorted[i];
    }
    preMargin /= marginSize - 1;
    for (let i = postSorted.length - marginSize; i < postSorted.length - 1; i++) {
      postMargin += postSorted[i + 1] - postSorted[i];
    }
    postMargin /= marginSize - 1;

    // Threshold agreement
    const thresholds = [0.5, 0.6, 0.7, 0.8, 0.9];
    const agreement: Record<number, number> = {};
    for (const tau of thresholds) {
      let agree = 0;
      for (let i = 0; i < post.length; i++) {
        const postDec = post[i] >= tau ? 1 : 0;
        const ceilDec = c[i] >= tau ? 1 : 0;
        if (postDec === ceilDec) agree++;
      }
      agreement[tau] = agree / post.length;
    }

    // Compute means
    let meanPre = 0;
    let meanPost = 0;
    let meanCeil = 0;
    for (let i = 0; i < m.length; i++) {
      meanPre += m[i];
      meanPost += post[i];
      meanCeil += c[i];
    }
    meanPre /= m.length;
    meanPost /= m.length;
    meanCeil /= c.length;

    this._report = {
      nPairs: m.length,
      meanMappedScore: meanPre,
      meanCeilingScore: meanCeil,
      meanShift: meanPost - meanPre,
      preMarginMean: preMargin,
      postMarginMean: postMargin,
      marginRatio: preMargin > 1e-12 ? postMargin / preMargin : 1.0,
      thresholdAgreement: agreement,
    };

    return this;
  }

  /**
   * Recalibrate scores: mapped -> ceiling-equivalent.
   */
  transform(scores: Float64Array | number[]): Float64Array {
    if (!this.isFitted || this._thresholds === null || this._values === null) {
      throw new Error('ScoreRecalibrator not fitted; call fit() first');
    }
    const input = Float64Array.from(scores);
    const result = new Float64Array(input.length);
    for (let i = 0; i < input.length; i++) {
      result[i] = interp(input[i], this._thresholds, this._values);
    }
    return result;
  }

  /**
   * Build recalibrator from holdout data.
   */
  static fitFromHoldout(
    docSrc: Float64Array[],
    docTgt: Float64Array[],
    qryTgt: Float64Array[],
    docIds: string[],
    queryIds: string[],
    qrels: Map<string, Set<string>>,
    mappedDocs: Float64Array[],
    options?: { nRandomPairs?: number; seed?: number },
  ): ScoreRecalibrator {
    const nRandomPairs = options?.nRandomPairs ?? 50_000;
    const seed = options?.seed ?? 0;
    const rng = createRng(seed);

    const nDocs = docIds.length;
    const nQueries = queryIds.length;

    const qtN = l2Normalize(qryTgt) as Float64Array[];
    const dtN = l2Normalize(docTgt) as Float64Array[];
    const dmN = l2Normalize(mappedDocs) as Float64Array[];

    const d = qtN[0].length;
    const mappedPairs: number[] = [];
    const ceilingPairs: number[] = [];

    // Sample random pairs
    const nRandom = Math.min(nRandomPairs, nQueries * nDocs);
    for (let i = 0; i < nRandom; i++) {
      const qi = Math.floor(rng() * nQueries);
      const di = Math.floor(rng() * nDocs);
      let mapDot = 0;
      let ceilDot = 0;
      for (let j = 0; j < d; j++) {
        mapDot += qtN[qi][j] * dmN[di][j];
        ceilDot += qtN[qi][j] * dtN[di][j];
      }
      mappedPairs.push(mapDot);
      ceilingPairs.push(ceilDot);
    }

    // Add top-20 neighbor pairs (cover the high-score region)
    const topK = Math.min(20, nDocs);
    for (let qi = 0; qi < nQueries; qi++) {
      // Compute all similarities for this query
      const sims = new Float64Array(nDocs);
      for (let di = 0; di < nDocs; di++) {
        let dot = 0;
        for (let j = 0; j < d; j++) {
          dot += qtN[qi][j] * dmN[di][j];
        }
        sims[di] = dot;
      }
      // Get top-K indices
      const indices = argSortDesc(sims).slice(0, topK);
      for (const di of indices) {
        let mapDot = 0;
        let ceilDot = 0;
        for (let j = 0; j < d; j++) {
          mapDot += qtN[qi][j] * dmN[di][j];
          ceilDot += qtN[qi][j] * dtN[di][j];
        }
        mappedPairs.push(mapDot);
        ceilingPairs.push(ceilDot);
      }
    }

    const recal = new ScoreRecalibrator();
    recal.fit(mappedPairs, ceilingPairs);
    return recal;
  }

  /**
   * Serialize to ScoreRecalibratorData for .isotrieve header storage.
   */
  toJSON(): ScoreRecalibratorData {
    if (!this.isFitted || this._thresholds === null || this._values === null) {
      throw new Error('Cannot serialize unfitted recalibrator');
    }
    return {
      thresholds: Array.from(this._thresholds),
      values: Array.from(this._values),
      report: this._report,
    };
  }

  /**
   * Load from serialized data.
   */
  static fromJSON(data: ScoreRecalibratorData): ScoreRecalibrator {
    const recal = new ScoreRecalibrator();
    recal._thresholds = new Float64Array(data.thresholds);
    recal._values = new Float64Array(data.values);
    recal._report = data.report ?? null;
    return recal;
  }
}

// ── Helpers ─────────────────────────────────────────────────────

function interp(x: number, xp: Float64Array, yp: Float64Array): number {
  const n = xp.length;
  if (n === 0) return 0;
  if (x <= xp[0]) return yp[0];
  if (x >= xp[n - 1]) return yp[n - 1];
  for (let i = 0; i < n - 1; i++) {
    if (x >= xp[i] && x <= xp[i + 1]) {
      const t = (x - xp[i]) / (xp[i + 1] - xp[i]);
      return yp[i] + t * (yp[i + 1] - yp[i]);
    }
  }
  return yp[n - 1];
}

function argSortDesc(arr: Float64Array): number[] {
  const indices = Array.from({ length: arr.length }, (_, i) => i);
  indices.sort((a, b) => arr[b] - arr[a]);
  return indices;
}

// Re-export createRng from math for internal use
function createRng(seed: number): () => number {
  let s = seed | 0;
  return () => {
    s = (s + 0x6d2b79f5) | 0;
    let t = Math.imul(s ^ (s >>> 15), 1 | s);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}
